docs/melange-for-x-developers.md

```diff
@@ -0,0 +1,601 @@
+```ocaml
+let add x y = x + y
+```
+
+### Algebraic data types
+
+It is not possible to build an ADT in TypeScript the same way as in OCaml.
+[Discriminated
+unions](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions)
+would be the closest analog to them, with libraries like
+[ts-pattern](https://github.com/gvergnaud/ts-pattern) as an alternative to the
+lack of support for pattern matching in the language.
+
+In OCaml, [algebraic data types
+(ADTs)](https://cs3110.github.io/textbook/chapters/data/algebraic_data_types.html)
+are a commonly used functionality of the language. They allow you to build your
+own types from small blocks. And with [pattern
+matching](https://ocaml.org/docs/data-types), it is easy to access this data.
+
+### Nominal typing
+
+In TypeScript, all typing is structural. This means that it is hard sometimes to
+establish a boundary or separation between two types that have the same
+implementation. For these cases, nominal typing can be emulated using tags:
+
+```js
+type Email = string & { readonly __tag: unique symbol };
+type City = string & { readonly __tag: unique symbol };
+```
+
+In OCaml, nominal typing is fully supported. Some of the core language types
+like [records and
+variants](https://v2.ocaml.org/manual/coreexamples.html#s%3Atut-recvariants) are
+nominal. This means that even if you declare exactly the same type twice,
+functions that operate on values from one type will not be compatible with the
+other type.
+
+There is also structural typing, used for OCaml
+[objects](https://v2.ocaml.org/manual/objectexamples.html) and [polymorphic
+variants](https://v2.ocaml.org/manual/polyvariant.html).
+
+### Immutability
+
+TypeScript has two base primitives to work with immutability: `const` and
+`readonly`.
+
+The first one is used to prevent variable reference change.
+
+```js
+const a = 1;
+a = 2; // Error: Cannot assign to 'a' because it is a constant.
+```
+
+And the second one is used to make properties immutable.
+
+```js
+type A = {
+  readonly x: number;
+}
+const a: A = { x: 1 };
+a.x = 12; // Error: Cannot assign to 'x' because it is a read-only property.
+```
+
+Nevertheless, there are some problems here. `const` and `readonly` only block
+reference changes but do nothing about values. With `const a = [1, 2, 3]` or
+`readonly x: number[]`, you can still change the contents of an array.
+
+OCaml provides data types with immutability in mind, like lists, records, or
+maps.
+
+### Strictness and soundness
+
+TypeScript you can use types like `any`, or other extensive types like
+`Function`. This can be mitigated using the `strict` option in the
+`tsconfig.json`. There is no such option in OCaml.
+
+TypeScript may sacrifice soundness for practicality when needed, like mentioned
+in [the
+handbook](https://www.typescriptlang.org/docs/handbook/type-compatibility.html#a-note-on-soundness).
+OCaml implementations provide methods like `Obj.magic` but they are more
```

Comment:
and the identity primitive, which we also documented.

Reply:
https://github.com/melange-re/melange-re.github.io/pull/39/commits/dde8ae32d450e5e5dff9c2837f6970ec7bf916d2